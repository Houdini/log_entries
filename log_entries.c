#include "postgres.h"

#include "catalog/pg_type.h"
#include "commands/trigger.h"
#include "executor/spi.h"
#include "miscadmin.h"
#include "utils/builtins.h"
#include "utils/nabstime.h"
#include "utils/rel.h"

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(log_entries);
#define MaxAttributes 2

Datum
log_entries(PG_FUNCTION_ARGS)
{
    Relation rel;
    TriggerData *trigger_data;
    Trigger *trigger;
    TupleDesc tupdesc;
    Datum current_user;
    HeapTuple rettuple;
    Datum newvals[MaxAttributes];
    int chattr[MaxAttributes];
 
    if (!CALLED_AS_TRIGGER(fcinfo))
        elog(ERROR, "log_entries should be called only as trigger");
    
    trigger_data = (TriggerData *) fcinfo->context;
    
    if (!TRIGGER_FIRED_FOR_ROW(trigger_data->tg_event))
        elog(ERROR, "log_entries should be fire only for row");
    
    if (TRIGGER_FIRED_BY_DELETE(trigger_data->tg_event))
        return (Datum) trigger_data->tg_trigtuple;
    
    trigger = trigger_data->tg_trigger;

    if (trigger->tgnargs != MaxAttributes)
        elog(ERROR, "log_entries, need two arguments");

    rel = trigger_data->tg_relation;
    tupdesc = rel->rd_att;
    
    if (SPI_gettypeid(tupdesc, SPI_fnumber(tupdesc, trigger->tgargs[0])) != ABSTIMEOID)
        elog(ERROR, "log_entries, first argument should be ABSTIME");

    if (SPI_gettypeid(tupdesc, SPI_fnumber(tupdesc, trigger->tgargs[1])) != TEXTOID)
        elog(ERROR, "log_entries, first argument should be TEXT");

    current_user = CStringGetTextDatum(GetUserNameFromId(GetUserId(), false));

    newvals[0] = GetCurrentAbsoluteTime();
    newvals[1] = current_user;
    chattr[0] = SPI_fnumber(tupdesc, trigger->tgargs[0]);
    chattr[1] = SPI_fnumber(tupdesc, trigger->tgargs[1]);

    rettuple = SPI_modifytuple(rel, trigger_data->tg_trigtuple, MaxAttributes, chattr, newvals, NULL);
    return PointerGetDatum(rettuple);
}