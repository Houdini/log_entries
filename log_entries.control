comment = 'Записываем кто и когда менял запись'
default_version = '1.0'
module_pathname = '$libdir/log_entries'
relocatable = true
