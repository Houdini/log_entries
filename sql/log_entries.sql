-- ===================================================================
-- create log_entries
-- ===================================================================
create extension log_entries;
create schema "log_entries_schema";
create table "log_entries_schema".test_table (
  id int not null,
  data text,
  updated_at abstime,
  updated_by text,
  constraint id_pkey primary key (id)
);

insert into "log_entries_schema".test_table values(1, '', null, 'text');
select id, updated_at, updated_by from "log_entries_schema".test_table;

-- log_entries should be called as trigger
select log_entries();

-- test trigger
create trigger log_entries
  before insert or delete or update on "log_entries_schema".test_table
  for each row execute procedure log_entries(updated_at, updated_by);

insert into "log_entries_schema".test_table values(2, null, null);
select id, updated_by from "log_entries_schema".test_table where id = 2;
select count(*) from "log_entries_schema".test_table where updated_at is not null and id = 2;

update "log_entries_schema".test_table set data = 'data' where id = 1;
select count(*) from "log_entries_schema".test_table where updated_at is null;
select count(*) from "log_entries_schema".test_table where id = 1 and updated_at is null;
delete from "log_entries_schema".test_table;
select * from "log_entries_schema".test_table;
