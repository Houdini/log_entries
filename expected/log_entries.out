-- ===================================================================
-- create log_entries
-- ===================================================================
create extension log_entries;
create schema "log_entries_schema";
create table "log_entries_schema".test_table (
  id int not null,
  data text,
  updated_at abstime,
  updated_by text,
  constraint id_pkey primary key (id)
);
insert into "log_entries_schema".test_table values(1, '', null, 'text');
select id, updated_at, updated_by from "log_entries_schema".test_table;
 id | updated_at | updated_by 
----+------------+------------
  1 |            | text
(1 row)

-- log_entries should be called as trigger
select log_entries();
ERROR:  log_entries should be called only as trigger
-- test trigger
create trigger log_entries
  before insert or delete or update on "log_entries_schema".test_table
  for each row execute procedure log_entries(updated_at, updated_by);
insert into "log_entries_schema".test_table values(2, null, null);
select id, updated_by from "log_entries_schema".test_table where id = 2;
 id | updated_by 
----+------------
  2 | houdini
(1 row)

select count(*) from "log_entries_schema".test_table where updated_at is not null and id = 2;
 count 
-------
     1
(1 row)

update "log_entries_schema".test_table set data = 'data' where id = 1;
select count(*) from "log_entries_schema".test_table where updated_at is null;
 count 
-------
     0
(1 row)

select count(*) from "log_entries_schema".test_table where id = 1 and updated_at is null;
 count 
-------
     0
(1 row)

delete from "log_entries_schema".test_table;
select * from "log_entries_schema".test_table;
 id | data | updated_at | updated_by 
----+------+------------+------------
(0 rows)

